### PR TITLE
Initial Gamepad support

### PR DESCRIPTION
**Pull request by est31**
Fixes most of #4153.

You can play minetest now with gamepads, but this PR does not indend to add full support for formspec control via gamepads.

### CODE REVIEW

File: src/client/joystick_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	// Now about the buttons simulated by the axes
+	JLO_A_PB(JoystickAxisCmb(KeyType::FORWARD,  1, -1, 1024));
+	JLO_A_PB(JoystickAxisCmb(KeyType::BACKWARD, 1,  1, 1024));
+	JLO_A_PB(JoystickAxisCmb(KeyType::LEFT,     0, -1, 1024));
+	JLO_A_PB(JoystickAxisCmb(KeyType::RIGHT,    0,  1, 1024));
+
+	return jlo;
+}
+
+JoystickLayout default_layout = create_default_layout();
```
Comment (Zeno-) [inline src/client/joystick_controller.cpp]: Should be static?

File: src/game.cpp [hunk 2]
```diff
@@ -1642,6 +1612,41 @@ class Game {
 	static void settingChangedCallback(const std::string &setting_name, void *data);
 	void readSettings();
 
+	bool getLeftClicked()
```
Comment (Zeno-) [inline src/game.cpp]: Can these be inlined?

Comment (est31) [inline src/game.cpp]: as in?

**Comment by Zeno-** [comment]
+1

File: src/client/joystick_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+	// The back button means "ESC".
+	JLO_B_PB(KeyType::ESC,        1 << 6,      1 << 6);
+
+	// The start button counts as modifier as well as use key.
+	// JLO_B_PB(KeyType::USE,        sb,          sb));
+
+	// Accessible without start modifier button pressed
+	// regardless whether four is pressed or not
+	JLO_B_PB(KeyType::SNEAK,      sb | 1 << 2, 1 << 2);
```
Comment (Zeno-) [inline src/client/joystick_controller.cpp]: Aren't there names somewhere for these bit positions? It'd be a lot nicer if they weren't magic values like bit position 2 and bit position 4 etc etc

Comment (est31) [inline src/client/joystick_controller.cpp]: Not really. The bit positions come directly from irrlicht, and irrlicht gets them from the OS probably.

Also, while the layout on most joysticks is generally the same, smaller things usually differ, like on my gamepad the four buttons are called 1,2,3,4 (bits 0 - 3), while on sofar's gamepad they are called XYAB. So even if giving these buttons names will just confuse people with different gamepads.

Also, assigning names to joystick button combinations is partly what the layout is about.

Comment (Zeno-) [inline src/client/joystick_controller.cpp]: Ok

**Comment by sofar** [comment]
Really good now.

- digging works (l1)
- placement/interact works (r1)
- movement and camera movement is good
- jump and sneak work

What's needed next:

- implement the 1024/-1024 threshold for moving - I still can't stand still
- make moving speed analog.
- make L2 and R2 cycle through the inventory items, so I can use all the items
- make B and Y useful:
    - B for inventory?
    - Y for "use" ?
- map "back" to escape?

Of course, formspecs are going to be *completely* unusable :)

**Comment by est31** [comment]
> * make moving speed analog

IDK, I think you already move slowly enough in minetest. Moving at the maximum value for my gamepad is a bit bad as it has a little bulge that most likely came from the molding process, so that it always scratches when I move the joystick to the maximum front.

> * map "back" to escape?

In theory this should be already done. IDK, does your back button have another number than mine? For me its number 6.

> * make B and Y useful:
>   *   B for inventory?
>   *   Y for "use" ?

B or (for me) 2 (aka the "red button"), maps to `SPECIAL_1`, which is required to travel fast if you are in fast mode but don't have fly. Its "E" on the keyboard per default.

Y or (for me) 4 (aka the "yellow button"), is used for button combinations. USE is not really implemented.

**Comment by est31** [comment]
Err sorry, of course, the use button **is** supported, its the same as `SPECIAL_1`.

**Comment by est31** [comment]
1e86c89f3614cf298916149a8f13d44ea671da64
2060fd9cbe587d7e8ffe0cecdd67925f13a56c05

**Pull request closed without merging**